pkg/koordlet/resmanager/resource_update_executor.go

```diff
@@ -135,3 +151,107 @@ func (rm *ResourceUpdateExecutor) needUpdate(currentResource ResourceUpdater) bo
 	}
 	return false
 }
+
+func NewLeveledResourceUpdateExecutor(name string, forceUpdateSeconds int) *LeveledResourceUpdateExecutor {
+	executor := &LeveledResourceUpdateExecutor{
+		ResourceUpdateExecutor: ResourceUpdateExecutor{
+			name:               name,
+			resourceCache:      cache.NewCacheDefault(),
+			forceUpdateSeconds: forceUpdateSeconds,
+			locker:             &sync.Mutex{},
+		},
+	}
+	return executor
+}
+
+// LeveledUpdateBatchByCache update a batch of resources by the level order cacheable. It firstly merge updates
+// resources from top to bottom, and then updates resources from bottom to top. It is compatible for some of resources
+// which just need to update once but not have an additional merge update.
+func (e *LeveledResourceUpdateExecutor) LeveledUpdateBatchByCache(resources [][]MergeableResourceUpdater) (updated bool) {
+	e.locker.Lock()
+	defer e.locker.Unlock()
+	var err error
+	for i := 0; i < len(resources); i++ {
+		for _, resource := range resources[i] {
+			if !e.needUpdate(resource) {
+				continue
+			}
+
+			if !resource.NeedMerge() {
+				err = resource.Update()
+			} else {
+				// NOTE: write merged resource into cache when the merge picks the old value
+				resource, err = resource.MergeUpdate()
+			}
+			if err != nil {
+				klog.Errorf("LeveledResourceUpdateExecutor merge update resource %v fail! error: %v",
```

Comment:
why not return the error?

Reply:
batch-update should not stop for any one resource.
e.g. an update failure on pod A with a transient container state should not disrupt updates on pod B